cs3api4lab/cs3apismanager.py

```diff
@@ -57,15 +140,374 @@ def save(self, model, path):
+
+        if content:
+            content = self._read_file(tmp_model.path)
+
+            if format is None:
+                format = "text"
+
+            if model['mimetype'] is None:
+                default_mime = {
+                    'text': 'text/plain',
+                    'base64': 'application/octet-stream'
+                }[format]
+                model['mimetype'] = default_mime
+
+            model.update(
+                content=content,
+                format=format,
+            )
+
+        return model
+
+    def _notebook_model(self, path, content):
+
+        model, tmp_model = self._create_base_model_from_cs3_container(path)
+        model['type'] = 'notebook'
+
+        if content:
+            file_content = self._read_file(tmp_model.path)
+            nb = nbformat.reads(file_content, as_version=4)
+            self.mark_trusted_cells(nb, path)
+            model['content'] = nb
+            model['format'] = 'json'
+            self.validate_notebook_model(model)
+
+        return model
+
+    def _is_dir(self, path):
+
+        if path == '/' or path == '' or path is None:
+            return True
+
+        parent_path = self._get_parent_path(path)
+        path = self._normalize_path(path)
+
+        try:
+            cs3_container = self._cs3_file_api().read_directory(parent_path, self.cs3_user_id, self.cs3_endpoint)
+        except Exception as ex:
+            self.log.error(u'Error while reading container: %s %s', path, ex, exc_info=True)
+            raise web.HTTPError(500, u'Unexpected error while reading container: %s %s' % (path, ex))
+
+        for cs3_model in cs3_container:
+
+            if cs3_model.type == self.TYPE_DIRECTORY and cs3_model.path == path:
+                return True
+            if cs3_model.type == self.TYPE_FILE and cs3_model.path == path:
+                return False
+
+        return False
+
+    def _create_base_model_from_cs3_container(self, path):
+
+        parent_path = self._get_parent_path(path)
+
+        cs3_file_api = self._cs3_file_api()
+        cs3_container = cs3_file_api.read_directory(parent_path, self.cs3_user_id, self.cs3_endpoint)
+
+        cs3_model = None
+        for cs3_tmp_model in cs3_container:
+            if cs3_tmp_model.type == self.TYPE_FILE and cs3_tmp_model.path == path:
+                cs3_model = cs3_tmp_model
+
+        if cs3_model is None:
+            raise web.HTTPError(404, u'%s is not a file' % path, reason='bad type')
+
+        model = self._convert_container_to_base_model(cs3_model.path, cs3_container)
+        return model, cs3_model
+
+    def _convert_container_to_base_model(self, path, cs3_container):
+
+        size = None
+        writable = False
+        created = datetime(1970, 1, 1, 0, 0, tzinfo=tz.UTC)
+        last_modified = datetime(1970, 1, 1, 0, 0, tzinfo=tz.UTC)
+
+        #
+        # Get data from container element
+        #
+        for cs3_model in cs3_container:
```

Comment:
The method should only receive the info of 1 file, and not a list of them, right?
And, if you found the path you wanted, you should `break`.